src/server/game/Entities/Player/Player.cpp

```diff
@@ -26712,3 +26768,237 @@ std::string Player::GetDebugInfo() const
+        TC_LOG_INFO("anticheat", "Player::CheckOnFlyHack :  Player has a MOVEMENTFLAG_SWIMMING, but not in water");
+
+        sWorld->SendGMText(LANG_GM_ANNOUNCE_AFK_SWIMMING, GetPlayerMovingMe()->GetName().c_str());
+        return false;
+    }
+    else
+    {
+        float z = GetMap()->GetHeight(GetPhaseMask(), npos.GetPositionX(), npos.GetPositionY(), pz + GetCollisionHeight() + 0.5f, true, 50.0f); // smart flyhacks -> SimpleFly
+        float diff = pz - z;
+        if (diff > 6.8f)
+            if (diff > 6.8f + GetHoverOffset()) // better calculate the second time for false situations, but not call GetHoverOffset everytime (economy resource)
+            {
+                float waterlevel = GetBaseMap()->GetWaterLevel(npos.GetPositionX(), npos.GetPositionY()); // water walking
+                if (waterlevel && waterlevel + GetCollisionHeight() + GetHoverOffset() > pz)
+                    return true;
+
+                float cx, cy, cz;
+                GetClosePoint(cx, cy, cz, DEFAULT_PLAYER_BOUNDING_RADIUS, 6.0f, 0, 6.8f); // first check
+                if (pz - cz > 6.8f)
+                {
+                    // check dynamic collision for transport (TODO navmesh for transport map)
+                    GetMap()->getObjectHitPos(GetPhaseMask(), GetPositionX(), GetPositionY(), GetPositionZ() + GetCollisionHeight(), cx, cy, cz + GetCollisionHeight(), cx, cy, cz, -GetCollisionHeight());
+
+                    if (pz - cz > 6.8f)
+                    {
+                        TC_LOG_INFO("anticheat", "Player::CheckOnFlyHack :  FlyHack Detected for Account id : %u, Player %s", GetPlayerMovingMe()->GetSession()->GetAccountId(), GetPlayerMovingMe()->GetName().c_str());
+                        TC_LOG_INFO("anticheat", "Player::========================================================");
+                        TC_LOG_INFO("anticheat", "Player::CheckOnFlyHack :  playerZ = %f", pz);
+                        TC_LOG_INFO("anticheat", "Player::CheckOnFlyHack :  normalZ = %f", z);
+                        TC_LOG_INFO("anticheat", "Player::CheckOnFlyHack :  checkz = %f", cz);
+                        sWorld->SendGMText(LANG_GM_ANNOUNCE_AFH, GetPlayerMovingMe()->GetName().c_str());
+                        return false;
+                    }
+                }
+            }
+    }
+
+    return true;
+}
+
+void Player::UpdateMovementInfo(MovementInfo const& movementInfo)
+{
+    SetLastMoveClientTimestamp(movementInfo.time);
+    SetLastMoveServerTimestamp(GameTime::GetGameTimeMS());
+}
+
+bool Player::CheckMovementInfo(MovementInfo const& movementInfo, bool jump)
+{
+    if (!sWorld->getBoolConfig(CONFIG_ANTICHEAT_SPEEDHACK_ENABLED))
+        return true;
+
+    uint32 ctime = GetLastMoveClientTimestamp();
+    if (ctime)
+    {
+        if (ToUnit()->IsFalling() || IsInFlight())
+            return true;
+
+        if (GetTransport() || GetVehicle() || GetVehicleKit())
```

Comment:
So if we mount on vehicle and use hack, can we ignore hack check here?

Reply:
Yes.We can get speed of vehicle? If yes, need to add it

Reply:
I think we can get speed of vehicle, but I'm not sure if here we can.
in WorldSession::HandleMovementOpcodes I think yes